### PR TITLE
Improve error message when slug is already in use

### PR DESCRIPTION
**Pull request by bduffany**
Also removes `FillGroup` from the userDB interface per discussions with @tylerwilliams

### CODE REVIEW

File: server/buildbuddy_server/buildbuddy_server.go [hunk 1]
```diff
@@ -128,14 +128,17 @@ func (s *BuildBuddyServer) GetGroup(ctx context.Context, req *grpb.GetGroupReque
...
 	}
-	group := &tables.Group{}
 	urlIdentifier := strings.TrimSpace(req.GetUrlIdentifier())
-	if urlIdentifier != "" {
-		group.URLIdentifier = &urlIdentifier
+	if urlIdentifier == "" {
+		return nil, status.InvalidArgumentError("URL identifier is required.")
 	}
-	if err := userDB.FillGroup(ctx, group); err != nil {
+	group, err := userDB.GetGroupByURLIdentifier(ctx, urlIdentifier)
```
Comment (tylerwilliams) [inline server/buildbuddy_server/buildbuddy_server.go:135]: It's a little atypical in go to return a nil value if err is nil, but it's not a rule.

You might consolidate this and just return status.NotFoundError("The requested organization was not found.") from inside GetGroupByUrlIdentifier. and then you don't need the if group == nil check here.

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: server/buildbuddy_server/buildbuddy_server.go (hunk 1 above)
Comment (bduffany) [reply server/buildbuddy_server/buildbuddy_server.go:135]: I don't want `GetGroupByURLIdentifier` to return an error in the case where we want to check whether an URL identifier is already taken -- in cases where it's not already taken, `nil` is returned and that's not really an error. I'm sort of using `nil` like an "Optional" value that is not present.

Comment (bduffany) [reply server/buildbuddy_server/buildbuddy_server.go:137]: Updated to check errors per our offline discussion

File: server/buildbuddy_server/buildbuddy_server.go [hunk 2]
```diff
@@ -223,9 +225,20 @@ func (s *BuildBuddyServer) CreateGroup(ctx context.Context, req *grpb.CreateGrou
 		OwnedDomain: groupOwnedDomain,
 	}
 	urlIdentifier := strings.TrimSpace(req.GetUrlIdentifier())
+
+	// Make sure the URL identifier isn't already taken by any existing group.
 	if urlIdentifier != "" {
+		existingGroup, err := userDB.GetGroupByURLIdentifier(ctx, urlIdentifier)
```
Comment (tylerwilliams) [inline server/buildbuddy_server/buildbuddy_server.go:231]: if existingGroup, err := userDB.GetGroupByURLIdentifier(ctx, urlIdentifier); existingGroup != nil {
  return status.InvalidArgumentError("URL is already in use")
} else {
  if err != notfound {
    return err
  }
}

how about this? just trying to map the code as closely as possible to the logic of like... if grp is found, return in use, else if other unexpected err, return it.

File: server/buildbuddy_server/buildbuddy_server.go [hunk 3]
```diff
@@ -248,15 +261,31 @@ func (s *BuildBuddyServer) UpdateGroup(ctx context.Context, req *grpb.UpdateGrou
...
-		GroupID: req.GetId(),
+	var group *tables.Group
+	var err error
+	urlIdentifier := strings.TrimSpace(req.GetUrlIdentifier())
+
+	// Make sure the new URL identifier isn't taken by a different group
+	// than the one being updated.
+	if urlIdentifier != "" {
+		group, err = userDB.GetGroupByURLIdentifier(ctx, urlIdentifier)
+		if err != nil && gstatus.Code(err) != gcodes.NotFound {
```
Comment (tylerwilliams) [inline server/buildbuddy_server/buildbuddy_server.go:272]: same as above?

File: server/buildbuddy_server/buildbuddy_server.go [hunk 4]
```diff
@@ -248,15 +261,31 @@ func (s *BuildBuddyServer) UpdateGroup(ctx context.Context, req *grpb.UpdateGrou
...
+			return nil, err
+		}
+		if group != nil && group.GroupID != req.GetId() {
+			return nil, status.InvalidArgumentError("URL is already in use")
+		}
 	}
-	if err := userDB.FillGroup(ctx, group); err != nil {
-		return nil, err
+
+	if group == nil {
```
Comment (tylerwilliams) [inline server/buildbuddy_server/buildbuddy_server.go:280]: I'm slightly confused by this logic -- feels like we should be doing:

if req.GetUrlIdent() {
  // read group by url ident
} else if req.GetId() {
  // read group by ID
} else {
  // some error case
}

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: server/buildbuddy_server/buildbuddy_server.go (hunk 4 above)
Comment (bduffany) [reply server/buildbuddy_server/buildbuddy_server.go:280]: Ah yep, that's much cleaner. Thanks!

File: server/buildbuddy_server/buildbuddy_server.go (hunk 2 above)
Comment (bduffany) [reply server/buildbuddy_server/buildbuddy_server.go:231]: Nice, much cleaner. Done!

File: server/buildbuddy_server/buildbuddy_server.go (hunk 3 above)
Comment (bduffany) [reply server/buildbuddy_server/buildbuddy_server.go:272]: Done!

**Review by tylerwilliams — APPROVED** [review]
LGTM

**Pull request merged**